lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp

```diff
@@ -0,0 +1,345 @@
+using UpdateRequest = SSSP::UpdateRequest;
+using UpdateRequestIndexer = SSSP::UpdateRequestIndexer;
+using SrcEdgeTile = SSSP::SrcEdgeTile;
+using SrcEdgeTileMaker = SSSP::SrcEdgeTileMaker;
+using SrcEdgeTilePushWrap = SSSP::SrcEdgeTilePushWrap;
+using ReqPushWrap = SSSP::ReqPushWrap;
+using OutEdgeRangeFn = SSSP::OutEdgeRangeFn;
+using TileRangeFn = SSSP::TileRangeFn;
+
+using PSchunk = katana::PerSocketChunkFIFO<kChunkSize>;
+using OBIM = katana::OrderedByIntegerMetric<UpdateRequestIndexer, PSchunk>;
+using OBIM_Barrier = katana::OrderedByIntegerMetric<
+    UpdateRequestIndexer, PSchunk>::with_barrier<true>::type;
+
+bool
+CheckReachability(Graph* graph, const GNode& source) {
+  katana::InsertBag<GNode> current_bag;
+  katana::InsertBag<GNode> next_bag;
+
+  current_bag.push(source);
+  graph->GetData<NodeCount>(source) = 1;
+
+  while (true) {
+    if (current_bag.begin() == current_bag.end())
+      break;
+
+    katana::do_all(
+        katana::iterate(current_bag),
+        [&](GNode n) {
+          for (auto edge : graph->edges(n)) {
+            auto dest = *(graph->GetEdgeDest(edge));
+            if (graph->GetData<NodeCount>(dest) == 0) {
+              graph->GetData<NodeCount>(dest) = 1;
+              next_bag.push(dest);
+            }
+          }
+        },
+        katana::steal());
+
+    current_bag.clear();
+    std::swap(current_bag, next_bag);
+  }
+
+  if (graph->GetData<NodeCount>(reportNode) == 0) {
+    return false;
+  }
+
+  katana::do_all(katana::iterate(*graph), [&graph](GNode n) {
+    graph->GetData<NodeCount>(n) = 0;
+  });
+
+  return true;
+}
+
+//delta stepping implementation for finding a shortest path from source to report node
+template <typename Item, typename OBIMTy, typename PushWrap, typename EdgeRange>
+void
+DeltaStepAlgo(
+    Graph* graph, const GNode& source, const PushWrap& pushWrap,
+    const EdgeRange& edgeRange,
+    katana::InsertBag<std::pair<uint32_t, Path*>>* bag,
+    katana::InsertBag<Path*>* path_pointers) {
+  //! [reducible for self-defined stats]
+  katana::GAccumulator<size_t> bad_work;
+  //! [reducible for self-defined stats]
+  katana::GAccumulator<size_t> wl_empty_work;
+
+  graph->GetData<NodeCount>(source) = 1;
+
+  katana::InsertBag<Item> init_bag;
+
+  Path* path = new Path();
+  path->last = NULL;
+  path_pointers->push(path);
+
+  pushWrap(init_bag, source, 0, path, "parallel");
+
+  katana::for_each(
+      katana::iterate(init_bag),
+      [&](const Item& item, auto& ctx) {
+        for (auto ii : edgeRange(item)) {
+          GNode dst = *(graph->GetEdgeDest(ii));
+          auto& ddata_count = graph->GetData<NodeCount>(dst);
+          auto& ddata_max = graph->GetData<NodeMax>(dst);
+
+          Distance ew = graph->GetEdgeData<EdgeWeight>(ii);
+          const Distance new_dist = item.distance + ew;
+
+          if ((ddata_count >= numPaths) && (ddata_max <= new_dist))
+            continue;
+
+          Path* path;
+          path = new Path();
```

Comment:
Concurrent allocation using `new` would be bad in performance. Use our custom allocators instead. Try something like:
```
class PathAlloc {
public:
  Path* NewPath() {
     Path* path = allocator_.allocate(1);
     allocator_.construct(path, Path());
     return path;
  }

  void DeletePath(Path* path) {
     allocator_.destroy(path);
     allocator_.deallocate(path, 1);
  }

private:
  katana::FixedSizeAllocator<Path> alloc_;
};
```

Then instead of calling `new Path()` and `delete (path)`, call `path_alloc.NewPath()` and `path_alloc.DeletePath(path)`. Declare `PathAlloc path_alloc;` before `for_each`.

Reply:
I am getting runtime errors with this approach. It seems it can't handle parallel/concurrent calls

Reply:
What error are you getting?

It definitely supports concurrent allocations. Check delaunay triangulation for an example usage.

Reply:
Thanks man. I will go through it then.

Reply:
I get this segmentation fault: `Thread 1 "k-shortest-path" received signal SIGSEGV, Segmentation fault.
katana::PerBackend::getRemote (this=0x55555592a4c0 <katana::getPTSBackend()::b>, thread=0, 
    offset=3968) at /home/udit/udit_katana_public_Jan5/libgalois/src/PerThreadStorage.cpp:155
155	  char* rbase = heads[thread].load(std::memory_order_relaxed);`

Reply:
I can help debug this. Can you please push your code?

Reply:
I just pushed it. Thank you.